New Aggregate Healthcheck

kovacshuni opened the pull request.

Zapadlo commented on aggregate-healthcheck-sidekick@.service:
```diff
@@ -7,7 +7,14 @@ After=aggregate-healthcheck@%i.service
-  etcdctl set /vulcand/frontends/aggregate-healthcheck-public/frontend '{\"Type\":\"http\", \"BackendId\":\"aggregate-healthcheck\", \"Route\":\"Path(`/__health`)\"}';"
+  etcdctl set /vulcand/frontends/aggregate-healthcheck-public/frontend '{\"Type\":\"http\", \"BackendId\":\"aggregate-healthcheck\", \"Route\":\"Path(`/__health`)\"}'; \
+  etcdctl set /vulcand/frontends/aggregate-healthcheck-public-gtg/frontend '{\"Type\":\"http\", \"BackendId\":\"aggregate-healthcheck\", \"Route\":\"Path(`/__gtg`)\"}'; \
+  etcdctl set /ft/healthcheck-categories/default/period_seconds 60; \
```
Unsure about the value of these, are they meant to be defaults that can be overwritten in etcd? They will reset every time service restarts though

Zapadlo commented on annotations-api-sidekick@.service:
```diff
@@ -14,7 +14,8 @@ ExecStart=/bin/sh -c "\
     export ANNOTATIONS_API_PORT=$(echo $(/usr/bin/docker port annotations-api-%i 8080) | cut -d':' -f2); \
-    etcdctl set /services/api-policy-component/%i/healthcheck /__health --ttl 60; \
+    etcdctl set /services/api-policy-component/%i/healthcheck/path /__health --ttl 60; \
+    etcdctl set /services/api-policy-component/%i/healthcheck/categories default,read --ttl 60; \
```
Does the category really need to be updated every 45s?

Zapadlo commented on organisation-read-api-sidekick@.service:
```diff
@@ -15,7 +15,8 @@ ExecStart=/bin/sh -c "\
     export ORGANISATION_READ_API_PORT=$(echo $(/usr/bin/docker port organisation-read-api-%i 8080) | cut -d':' -f2); \
-    etcdctl set /services/organisation-read-api/%i/healthcheck /__health --ttl 60; \
+    etcdctl set /services/organisation-read-api/%i/healthcheck/path /__health --ttl 60; \
+    etcdctl set /services/organisation-read-api/%i/healthcheck/categories default,read --ttl 60; \
```
Again, inside the loop

Zapadlo commented on people-read-api-sidekick@.service:
```diff
@@ -15,7 +15,8 @@ ExecStart=/bin/sh -c "\
     export PEOPLE_READ_API_PORT=$(echo $(/usr/bin/docker port people-read-api-%i 8080) | cut -d':' -f2); \
-    etcdctl set /services/people-read-api/%i/healthcheck /__health --ttl 60; \
+    etcdctl set /services/people-read-api/%i/healthcheck/path /__health --ttl 60; \
+    etcdctl set /services/people-read-api/%i/healthcheck/categories default,read --ttl 60; \
```
Here

Zapadlo commented on semantic-reader-sidekick@.service:
```diff
@@ -17,7 +17,8 @@ ExecStart=/bin/sh -c "\
     etcdctl set /services/semantic-reader/%i/admin_port $(echo $(/usr/bin/docker port semantic-reader-%i 8081) | cut -d':' -f2) --ttl 60; \
-    etcdctl set /services/semantic-reader/%i/healthcheck /__health --ttl 60; \
+    etcdctl set /services/semantic-reader/%i/healthcheck/path /__health --ttl 60; \
+    etcdctl set /services/semantic-reader/%i/healthcheck/categories default --ttl 60; \
```
Here

Zapadlo commented on wp-notifier-sidekick@.service:
```diff
@@ -16,7 +16,8 @@ ExecStart=/bin/sh -c "\
     etcdctl set /services/wp-notifier/%i/admin_port $(echo $(/usr/bin/docker port wp-notifier-%i 8081) | cut -d':' -f2) --ttl 60; \
-    etcdctl set /services/wp-notifier/%i/healthcheck /__health --ttl 60; \
+    etcdctl set /services/wp-notifier/%i/healthcheck/path /__health --ttl 60; \
+    etcdctl set /services/wp-notifier/%i/healthcheck/categories default --ttl 60; \
```
Here

Zapadlo commented on wp-transformer-sidekick@.service:
```diff
@@ -16,7 +16,8 @@ ExecStart=/bin/sh -c "\
     etcdctl set /services/wp-transformer/%i/admin_port $(echo $(/usr/bin/docker port wp-transformer-%i 8081) | cut -d':' -f2) --ttl 60; \
-    etcdctl set /services/wp-transformer/%i/healthcheck /__health --ttl 60; \
+    etcdctl set /services/wp-transformer/%i/healthcheck/path /__health --ttl 60; \
+    etcdctl set /services/wp-transformer/%i/healthcheck/categories default --ttl 60; \
```
And here

catalin-ilea commented on api-policy-component-sidekick@.service:
```diff
@@ -7,7 +7,8 @@ After=api-policy-component@%i.service
   etcdctl set /ft/services/api-policy-component/healthcheck true; \
   etcdctl mkdir /ft/services/api-policy-component/servers; \
-  etcdctl set /ft/healthcheck/api-policy-component-%i /__health; \
+  etcdctl set /ft/healthcheck/api-policy-component-%i/path /__health; \
```
Isn't this redundant? Is the health path ever going to be different?

catalin-ilea commented on binary-ingester-sidekick@.service:
```diff
@@ -14,11 +13,14 @@ ExecStart=/bin/sh -c "\
-  etcdctl set /ft/healthcheck/binary-ingester-%i /__health; \
-  \
+  etcdctl set /ft/healthcheck/binary-ingester-%i/path /__health; \
+  etcdctl set /ft/healthcheck/binary-ingester-%i/categories default; \
```
Should we assume default if the key is missing? This way you'd avoid a big-bang release where all services will get restarted.

Zapadlo replied:
+1

Zapadlo replied:
+1

kovacshuni commented:
Okay, so we talked about the etcd keys with Catalin.
I'm removing categories where not necessary, the absence will mean default.
We tried to do something to not need the modification for the path either but that can't be done, as the key /ft/healthcheck/service-name-1 is now a directory and before it was a leaf node. Categories wouldn't have a place to be in if it stayed a leaf node. So remained a big-bang-ish release for now. Hope it goes well.

kovacshuni commented:
I updated.
Default category is not even defined, and not assigned, it will be part of the agg-health program to associate with all services and its time will be 60 seconds from code and non-resilient.
Extracted the lines from while loops to the outside.
The .../path __health i had to leave, that marks a service that it is defined and has to be checked upon.

tosan88 commented:
:+1:

Zapadlo commented:
+1

catalin-ilea commented:
Can you please double check you modified all services which are deployed and expose HCs? You're missing topics-rw-neo4j...might be others as well

kovacshuni commented:
I merged latest master in here. Fixed up topics-rw-neo4j, this was a new one. Double checked all services have their etcd health key correct. Testin in xp.

catalin-ilea commented:
Brill

kovacshuni commented:
```
Agg-hc deployment plan

#clean-up old etcd values, right after merge

1. etcdctl rm --recursive /ft/healthcheck
2. fleetctl list-units | grep sidekick | awk '{print $1}' | xargs fleetctl destroy
```

kovacshuni commented:
* Tested in XP - worked like a charm
* Tested in pre-prod. - worked like a charm
